File: ois/ois20/round3/ois11_round3_solutions.tar(1)/scmax.pas
{ NOTE: it is recommended to use this even if you don't understand the following code. }

{ constraints }
const
    MAXN = 100000;

{ input data }
var
    N, i, temp : longint;
    A     : array[0..MAXN-1] of longint;
    T     : array[1..MAXN] of longint;
    ft    : array[1..MAXN] of longint;
    best    : array[1..MAXN] of longint;

procedure update(i: longint; v: longint);
begin
    i := N-i+1;
    while (i <= N) do
    begin
        if v > ft[i] then
            ft[i] := v;
        i := i + (i and (-i));
    end;
end;

function get(i:longint) : longint;
var res: longint;
begin
    res := 0;
    i := N-i+1;
    while (i > 0) do
    begin
        if res < ft[i] then
            res := ft[i];
        i := i - (i and (-i));
    end;
    get := res;
end;


begin
{
    uncomment the following lines if you want to read/write from files
    assign(input,  'input.txt');  reset(input);
    assign(output, 'output.txt'); rewrite(output);
}

    readln(N);
    for i:=0 to N-1 do
        read(A[i]);
    readln();
    for i:=1 to N do
        read(T[i]);
    readln();

    {
        WARNING! T is indexed from 1!
        In particular T[i] is the preferred number of i.
    }

    for i:=1 to N do
    begin
        best[i] := 0;
        ft[i] := 0;
    end;

    i := N-1;
    while i >= 0 do
    begin
        if best[T[A[i]]]+1 > best[A[i]] then
            best[A[i]] := best[T[A[i]]]+1;
        temp := get(A[i]+1)+1;
        if temp > best[A[i]] then
            best[A[i]] := temp;
        update(A[i], best[A[i]]);
        i := i-1;
    end;



    { insert your code here }

    writeln(get(1)); { print result }
end.

File: ois/ois20/round3/ois11_round3_solutions.tar(1)/scmax.cpp
/*
 * This template is valid both in C and in C++,
 * so you can expand it with code from both languages.
 * NOTE: it is recommended to use this even if you don't
 * understand the following code.
 */

#include <stdio.h>
#include <assert.h>
#include <algorithm>

// constraints
#define MAXN 100000

using namespace std;

// input data
int N, i;
int A[MAXN], T[MAXN+1];
int best[MAXN+1];
int tree[4*MAXN];

int get(int a, int b, int l, int r, int n) {
    if (b < l || a > r) {
        return 0;
    }

    if (l >= a && r <= b) {
        return tree[n];
    }

    return max(get(a, b, l, (l+r)/2, n*2), get(a, b, (l+r)/2+1, r, n*2+1));
}

int get(int i) {
    return get(i, N, 1, N, 1);
}

void update(int i, int v, int l, int r, int n) {
    if (i < l || i > r) return;
    if (i == l && l == r) {
        tree[n] = v;
        return;
    }
    update(i, v, l, (l+r)/2, n*2);
    update(i, v, (l+r)/2+1, r, n*2+1);
    tree[n] = max(tree[n*2], tree[n*2+1]);
}

void update(int i, int v) {
    update(i, v, 1, N, 1);
}

int main() {
//  uncomment the following lines if you want to read/write from files
//  freopen("input.txt", "r", stdin);
//  freopen("output.txt", "w", stdout);

    assert(1 == scanf("%d", &N));
    for(i=0; i<N; i++)
        assert(1 == scanf("%d", &A[i]));

    for(i=1; i<=N; i++)
        assert(1 == scanf("%d", &T[i]));

    // WARNING! T is indexed from 1!
    // In particular T[i] is the preferred number of i.

    // insert your code here

    for (int i = N-1; i >= 0; i--) {
        best[A[i]] = max(best[A[i]], best[T[A[i]]]+1);
        best[A[i]] = max(best[A[i]], get(A[i]+1)+1);
        update(A[i], best[A[i]]);
    }

    printf("%d\n", get(1)); // print the result
    return 0;
}
